// @deck.gl/aggregation-layers/dist/es5/grid-aggregation-layer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _aggregationLayer = _interopRequireDefault(require("./aggregation-layer"));

var _gpuGridAggregator = _interopRequireDefault(require("./utils/gpu-grid-aggregation/gpu-grid-aggregator"));

var _core = require("@luma.gl/core");

var _core2 = require("@deck.gl/core");

var _binSorter = _interopRequireDefault(require("./utils/bin-sorter"));

var _gridAggregator = require("./cpu-grid-layer/grid-aggregator");

class GridAggregationLayer extends _aggregationLayer.default {
  initializeState({
    dimensions
  }) {
    const {
      gl
    } = this.context;
    super.initializeState(dimensions);
    this.setState({
      layerData: {},
      gpuGridAggregator: new _gpuGridAggregator.default(gl, {
        id: "".concat(this.id, "-gpu-aggregator")
      }),
      cpuGridAggregator: _gridAggregator.pointToDensityGridDataCPU
    });
  }

  updateState(opts) {
    super.updateState(opts);
    this.updateAggregationState(opts);
    const {
      aggregationDataDirty,
      aggregationWeightsDirty,
      gpuAggregation
    } = this.state;

    if (this.getNumInstances() <= 0) {
      return;
    }

    let aggregationDirty = false;

    if (aggregationDataDirty || gpuAggregation && aggregationWeightsDirty) {
      this._updateAggregation(opts);

      aggregationDirty = true;
    }

    if (!gpuAggregation && (aggregationDataDirty || aggregationWeightsDirty)) {
      this._updateWeightBins();

      this._uploadAggregationResults();

      aggregationDirty = true;
    }

    this.setState({
      aggregationDirty
    });
  }

  finalizeState() {
    var _this$state$gpuGridAg;

    const {
      count
    } = this.state.weights;

    if (count && count.aggregationBuffer) {
      count.aggregationBuffer.delete();
    }

    (_this$state$gpuGridAg = this.state.gpuGridAggregator) === null || _this$state$gpuGridAg === void 0 ? void 0 : _this$state$gpuGridAg.delete();
    super.finalizeState();
  }

  updateShaders(shaders) {
    if (this.state.gpuAggregation) {
      this.state.gpuGridAggregator.updateShaders(shaders);
    }
  }

  updateAggregationState(opts) {
    _core2.log.assert(false);
  }

  allocateResources(numRow, numCol) {
    if (this.state.numRow !== numRow || this.state.numCol !== numCol) {
      const dataBytes = numCol * numRow * 4 * 4;
      const gl = this.context.gl;
      const {
        weights
      } = this.state;

      for (const name in weights) {
        const weight = weights[name];

        if (weight.aggregationBuffer) {
          weight.aggregationBuffer.delete();
        }

        weight.aggregationBuffer = new _core.Buffer(gl, {
          byteLength: dataBytes,
          accessor: {
            size: 4,
            type: 5126,
            divisor: 1
          }
        });
      }
    }
  }

  updateResults({
    aggregationData,
    maxMinData,
    maxData,
    minData
  }) {
    const {
      count
    } = this.state.weights;

    if (count) {
      count.aggregationData = aggregationData;
      count.maxMinData = maxMinData;
      count.maxData = maxData;
      count.minData = minData;
    }
  }

  _updateAggregation(opts) {
    const {
      cpuGridAggregator,
      gpuGridAggregator,
      gridOffset,
      posOffset,
      translation = [0, 0],
      scaling = [0, 0, 0],
      boundingBox,
      projectPoints,
      gpuAggregation,
      numCol,
      numRow
    } = this.state;
    const {
      props
    } = opts;
    const {
      viewport
    } = this.context;
    const attributes = this.getAttributes();
    const vertexCount = this.getNumInstances();

    if (!gpuAggregation) {
      const result = cpuGridAggregator(props, {
        gridOffset,
        projectPoints,
        attributes,
        viewport,
        posOffset,
        boundingBox
      });
      this.setState({
        layerData: result
      });
    } else {
      const {
        weights
      } = this.state;
      gpuGridAggregator.run({
        weights,
        cellSize: [gridOffset.xOffset, gridOffset.yOffset],
        numCol,
        numRow,
        translation,
        scaling,
        vertexCount,
        projectPoints,
        attributes,
        moduleSettings: this.getModuleSettings()
      });
    }
  }

  _updateWeightBins() {
    const {
      getValue
    } = this.state;
    const sortedBins = new _binSorter.default(this.state.layerData.data || [], {
      getValue
    });
    this.setState({
      sortedBins
    });
  }

  _uploadAggregationResults() {
    const {
      numCol,
      numRow
    } = this.state;
    const {
      data
    } = this.state.layerData;
    const {
      aggregatedBins,
      minValue,
      maxValue,
      totalCount
    } = this.state.sortedBins;
    const ELEMENTCOUNT = 4;
    const aggregationSize = numCol * numRow * ELEMENTCOUNT;
    const aggregationData = new Float32Array(aggregationSize).fill(0);

    for (const bin of aggregatedBins) {
      const {
        lonIdx,
        latIdx
      } = data[bin.i];
      const {
        value,
        counts
      } = bin;
      const cellIndex = (lonIdx + latIdx * numCol) * ELEMENTCOUNT;
      aggregationData[cellIndex] = value;
      aggregationData[cellIndex + ELEMENTCOUNT - 1] = counts;
    }

    const maxMinData = new Float32Array([maxValue, 0, 0, minValue]);
    const maxData = new Float32Array([maxValue, 0, 0, totalCount]);
    const minData = new Float32Array([minValue, 0, 0, totalCount]);
    this.updateResults({
      aggregationData,
      maxMinData,
      maxData,
      minData
    });
  }

}

exports.default = GridAggregationLayer;
GridAggregationLayer.layerName = 'GridAggregationLayer';
