// @deck.gl/aggregation-layers/dist/es5/utils/gpu-grid-aggregation/gpu-grid-aggregator-constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MAX_MIN_TEXTURE_OPTS = exports.WEIGHT_SIZE = exports.PIXEL_SIZE = exports.DEFAULT_WEIGHT_PARAMS = exports.ELEMENTCOUNT = exports.EQUATION_MAP = exports.MAX_MIN_BLEND_EQUATION = exports.MAX_BLEND_EQUATION = exports.MIN_BLEND_EQUATION = exports.MAX_32_BIT_FLOAT = exports.DEFAULT_RUN_PARAMS = void 0;

var _aggregationOperationUtils = require("../aggregation-operation-utils");

const DEFAULT_RUN_PARAMS = {
  projectPoints: false,
  viewport: null,
  createBufferObjects: true,
  moduleSettings: {}
};
exports.DEFAULT_RUN_PARAMS = DEFAULT_RUN_PARAMS;
const MAX_32_BIT_FLOAT = 3.402823466e38;
exports.MAX_32_BIT_FLOAT = MAX_32_BIT_FLOAT;
const MIN_BLEND_EQUATION = [32775, 32774];
exports.MIN_BLEND_EQUATION = MIN_BLEND_EQUATION;
const MAX_BLEND_EQUATION = [32776, 32774];
exports.MAX_BLEND_EQUATION = MAX_BLEND_EQUATION;
const MAX_MIN_BLEND_EQUATION = [32776, 32775];
exports.MAX_MIN_BLEND_EQUATION = MAX_MIN_BLEND_EQUATION;
const EQUATION_MAP = {
  [_aggregationOperationUtils.AGGREGATION_OPERATION.SUM]: 32774,
  [_aggregationOperationUtils.AGGREGATION_OPERATION.MEAN]: 32774,
  [_aggregationOperationUtils.AGGREGATION_OPERATION.MIN]: MIN_BLEND_EQUATION,
  [_aggregationOperationUtils.AGGREGATION_OPERATION.MAX]: MAX_BLEND_EQUATION
};
exports.EQUATION_MAP = EQUATION_MAP;
const ELEMENTCOUNT = 4;
exports.ELEMENTCOUNT = ELEMENTCOUNT;
const DEFAULT_WEIGHT_PARAMS = {
  size: 1,
  operation: _aggregationOperationUtils.AGGREGATION_OPERATION.SUM,
  needMin: false,
  needMax: false,
  combineMaxMin: false
};
exports.DEFAULT_WEIGHT_PARAMS = DEFAULT_WEIGHT_PARAMS;
const PIXEL_SIZE = 4;
exports.PIXEL_SIZE = PIXEL_SIZE;
const WEIGHT_SIZE = 3;
exports.WEIGHT_SIZE = WEIGHT_SIZE;
const MAX_MIN_TEXTURE_OPTS = {
  format: 34836,
  type: 5126,
  border: 0,
  mipmaps: false,
  parameters: {
    [10240]: 9728,
    [10241]: 9728
  },
  dataFormat: 6408,
  width: 1,
  height: 1
};
exports.MAX_MIN_TEXTURE_OPTS = MAX_MIN_TEXTURE_OPTS;
