// @deck.gl/json/dist/es5/json-configuration.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _parseExpressionString = _interopRequireDefault(require("./helpers/parse-expression-string"));

var _assert = _interopRequireDefault(require("./utils/assert"));

var _syntacticSugar = require("./syntactic-sugar");

const isObject = value => value && typeof value === 'object';

class JSONConfiguration {
  constructor(...configurations) {
    this.typeKey = _syntacticSugar.TYPE_KEY;
    this.functionKey = _syntacticSugar.FUNCTION_KEY;
    this.log = console;
    this.classes = {};
    this.reactComponents = {};
    this.enumerations = {};
    this.constants = {};
    this.functions = {};
    this.convertFunction = _parseExpressionString.default;

    this.preProcessClassProps = (Class, props) => props;

    this.postProcessConvertedJson = json => json;

    for (const configuration of configurations) {
      this.merge(configuration);
    }
  }

  merge(configuration) {
    for (const key in configuration) {
      switch (key) {
        case 'layers':
        case 'views':
          Object.assign(this.classes, configuration[key]);
          break;

        default:
          if (key in this) {
            const value = configuration[key];
            this[key] = isObject(this[key]) ? Object.assign(this[key], value) : value;
          }

      }
    }
  }

  validate(configuration) {
    (0, _assert.default)(!this.typeKey || typeof this.typeKey === 'string');
    (0, _assert.default)(isObject(this.classes));
    return true;
  }

}

exports.default = JSONConfiguration;
