// @loaders.gl/core/dist/es5/lib/loader-utils/loggers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConsoleLog = exports.NullLog = exports.probeLog = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _probe = require("probe.gl");

const probeLog = new _probe.Log({
  id: 'loaders.gl'
});
exports.probeLog = probeLog;

class NullLog {
  log() {
    return () => {};
  }

  info() {
    return () => {};
  }

  warn() {
    return () => {};
  }

  error() {
    return () => {};
  }

}

exports.NullLog = NullLog;

class ConsoleLog {
  constructor() {
    (0, _defineProperty2.default)(this, "console", void 0);
    this.console = console;
  }

  log(...args) {
    return this.console.log.bind(this.console, ...args);
  }

  info(...args) {
    return this.console.info.bind(this.console, ...args);
  }

  warn(...args) {
    return this.console.warn.bind(this.console, ...args);
  }

  error(...args) {
    return this.console.error.bind(this.console, ...args);
  }

}

exports.ConsoleLog = ConsoleLog;
