// @loaders.gl/tiles/dist/es5/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TILE3D_OPTIMIZATION_HINT = exports.LOD_METRIC_TYPE = exports.TILESET_TYPE = exports.TILE_TYPE = exports.TILE_REFINEMENT = exports.TILE_CONTENT_STATE = void 0;
const TILE_CONTENT_STATE = {
  UNLOADED: 0,
  LOADING: 1,
  PROCESSING: 2,
  READY: 3,
  EXPIRED: 4,
  FAILED: 5
};
exports.TILE_CONTENT_STATE = TILE_CONTENT_STATE;
const TILE_REFINEMENT = {
  ADD: 1,
  REPLACE: 2
};
exports.TILE_REFINEMENT = TILE_REFINEMENT;
const TILE_TYPE = {
  EMPTY: 'empty',
  SCENEGRAPH: 'scenegraph',
  POINTCLOUD: 'pointcloud',
  MESH: 'mesh'
};
exports.TILE_TYPE = TILE_TYPE;
const TILESET_TYPE = {
  I3S: 'I3S',
  TILES3D: 'TILES3D'
};
exports.TILESET_TYPE = TILESET_TYPE;
const LOD_METRIC_TYPE = {
  GEOMETRIC_ERROR: 'geometricError',
  MAX_SCREEN_THRESHOLD: 'maxScreenThreshold'
};
exports.LOD_METRIC_TYPE = LOD_METRIC_TYPE;
const TILE3D_OPTIMIZATION_HINT = {
  NOT_COMPUTED: -1,
  USE_OPTIMIZATION: 1,
  SKIP_OPTIMIZATION: 0
};
exports.TILE3D_OPTIMIZATION_HINT = TILE3D_OPTIMIZATION_HINT;
