// @loaders.gl/images/dist/es5/image-loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports._typecheckImageLoader = exports.ImageLoader = void 0;

var _version = require("./lib/utils/version");

var _parseImage = _interopRequireDefault(require("./lib/parsers/parse-image"));

var _binaryImageApi = require("./lib/category-api/binary-image-api");

const EXTENSIONS = ['png', 'jpg', 'jpeg', 'gif', 'webp', 'bmp', 'ico', 'svg'];
const MIME_TYPES = ['image/png', 'image/jpeg', 'image/gif', 'image/webp', 'image/bmp', 'image/vnd.microsoft.icon', 'image/svg+xml'];
const ImageLoader = {
  id: 'image',
  module: 'images',
  name: 'Images',
  version: _version.VERSION,
  mimeTypes: MIME_TYPES,
  extensions: EXTENSIONS,
  parse: _parseImage.default,
  tests: [arrayBuffer => Boolean((0, _binaryImageApi.getBinaryImageMetadata)(new DataView(arrayBuffer)))],
  options: {
    image: {
      type: 'auto',
      decode: true
    }
  }
};
exports.ImageLoader = ImageLoader;
const _typecheckImageLoader = ImageLoader;
exports._typecheckImageLoader = _typecheckImageLoader;
