// @deck.gl/extensions/dist/es5/fp64/project64.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _shadertools = require("@luma.gl/shadertools");

var _core = require("@deck.gl/core");

var _project = _interopRequireDefault(require("./project64.glsl"));

const {
  fp64ify,
  fp64ifyMatrix4
} = _shadertools.fp64;
var _default = {
  name: 'project64',
  dependencies: [_core.project, _shadertools.fp64],
  vs: _project.default,
  getUniforms
};
exports.default = _default;
const DEFAULT_MODULE_OPTIONS = {};
const getMemoizedUniforms = (0, _core._memoize)(calculateUniforms);

function getUniforms(opts = DEFAULT_MODULE_OPTIONS) {
  if (opts.viewport) {
    const {
      viewProjectionMatrix,
      scale
    } = opts.viewport;
    return getMemoizedUniforms({
      viewProjectionMatrix,
      scale
    });
  }

  return {};
}

function calculateUniforms({
  viewProjectionMatrix,
  scale
}) {
  const glViewProjectionMatrixFP64 = fp64ifyMatrix4(viewProjectionMatrix);
  const scaleFP64 = fp64ify(scale);
  return {
    project_uViewProjectionMatrixFP64: glViewProjectionMatrixFP64,
    project64_uViewProjectionMatrix: glViewProjectionMatrixFP64,
    project64_uScale: scaleFP64
  };
}
