// @deck.gl/carto/dist/es5/api/maps-api-common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.encodeParameter = encodeParameter;
exports.FORMATS = exports.MAP_TYPES = exports.DEFAULT_MAPS_URL_FORMAT = exports.API_VERSIONS = exports.DEFAULT_REGION_COMPONENT_IN_URL = exports.DEFAULT_USER_COMPONENT_IN_URL = void 0;
const DEFAULT_USER_COMPONENT_IN_URL = '{user}';
exports.DEFAULT_USER_COMPONENT_IN_URL = DEFAULT_USER_COMPONENT_IN_URL;
const DEFAULT_REGION_COMPONENT_IN_URL = '{region}';
exports.DEFAULT_REGION_COMPONENT_IN_URL = DEFAULT_REGION_COMPONENT_IN_URL;
const API_VERSIONS = {
  V1: 'v1',
  V2: 'v2',
  V3: 'v3'
};
exports.API_VERSIONS = API_VERSIONS;
const DEFAULT_MAPS_URL_FORMAT = {
  [API_VERSIONS.V1]: "https://".concat(DEFAULT_USER_COMPONENT_IN_URL, ".carto.com/api/v1/map"),
  [API_VERSIONS.V2]: "https://maps-api-v2.".concat(DEFAULT_REGION_COMPONENT_IN_URL, ".carto.com/user/").concat(DEFAULT_USER_COMPONENT_IN_URL)
};
exports.DEFAULT_MAPS_URL_FORMAT = DEFAULT_MAPS_URL_FORMAT;
const MAP_TYPES = {
  QUERY: 'query',
  TABLE: 'table',
  TILESET: 'tileset'
};
exports.MAP_TYPES = MAP_TYPES;
const FORMATS = {
  GEOJSON: 'geojson',
  NDJSON: 'ndjson',
  TILEJSON: 'tilejson',
  JSON: 'json'
};
exports.FORMATS = FORMATS;

function encodeParameter(name, value) {
  return "".concat(name, "=").concat(encodeURIComponent(value));
}
