// @deck.gl/core/dist/es5/transitions/viewport-fly-to-interpolator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _transitionInterpolator = _interopRequireDefault(require("./transition-interpolator"));

var _math = require("math.gl");

var _webMercator = require("@math.gl/web-mercator");

const LINEARLY_INTERPOLATED_PROPS = ['bearing', 'pitch'];
const DEFAULT_OPTS = {
  speed: 1.2,
  curve: 1.414
};

class FlyToInterpolator extends _transitionInterpolator.default {
  constructor(props = {}) {
    super({
      compare: ['longitude', 'latitude', 'zoom', 'bearing', 'pitch'],
      extract: ['width', 'height', 'longitude', 'latitude', 'zoom', 'bearing', 'pitch'],
      required: ['width', 'height', 'latitude', 'longitude', 'zoom']
    });
    this.props = { ...DEFAULT_OPTS,
      ...props
    };
  }

  interpolateProps(startProps, endProps, t) {
    const viewport = (0, _webMercator.flyToViewport)(startProps, endProps, t, this.props);

    for (const key of LINEARLY_INTERPOLATED_PROPS) {
      viewport[key] = (0, _math.lerp)(startProps[key] || 0, endProps[key] || 0, t);
    }

    return viewport;
  }

  getDuration(startProps, endProps) {
    let {
      transitionDuration
    } = endProps;

    if (transitionDuration === 'auto') {
      transitionDuration = (0, _webMercator.getFlyToDuration)(startProps, endProps, this.props);
    }

    return transitionDuration;
  }

}

exports.default = FlyToInterpolator;
